File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.ans = 0

        def depth(root):
            if not root:
                return 0
            ld = depth(root.left)
            rd = depth(root.right)
            self.ans = max(self.ans, ld + rd)

            return 1 + max(ld, rd)
        depth(root)
        return self.ans

File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        int left=1;
        int right=n-2;
        int lMax=height[left-1];
        int rMax=height[right+1];
        int res=0;
        while(left<=right){
            if(rMax<=lMax){
                res += max(0, rMax-height[right]);
                rMax = max(rMax, height[right]);
                right--;
            }
            else{
                res += max(0, lMax-height[left]);
                lMax = max(lMax, height[left]);
                left++;
            }
        }
        return res;
    }
};
